Vegeting/Global/Extensions/UIImage+Extension.swift

```diff
@@ -0,0 +1,32 @@
+//
+//  UIImage_Extension.swift
+//  Vegeting
+//
+//  Created by kelly on 2022/11/15.
+//
+
+import UIKit
+
+extension UIImage {
+    // 해상도를 너무 높지 않게 조절, 파일 크기가 너무 높지 않도록
+    var scaledToSafeUploadSize: UIImage? {
+        let maxImageSideLength: CGFloat = 480
+        let largerSide: CGFloat = max(size.width, size.height)
+        let ratioScale: CGFloat = largerSide > maxImageSideLength ? largerSide / maxImageSideLength : 1
+        let newImageSize = CGSize(width: size.width / ratioScale, height: size.height / ratioScale)
+        
+        return image(scaledTo: newImageSize)
+    }
+    
+    // 이미지를 입력받은 size에 맞춰 조절
+    private func image(scaledTo size: CGSize) -> UIImage? {
+        defer {
+            UIGraphicsEndImageContext()
+        }
+        
+        UIGraphicsBeginImageContextWithOptions(size, true, 0)
+        draw(in: CGRect(origin: .zero, size: size))
+        
+        return UIGraphicsGetImageFromCurrentImageContext()
+    }
```

Comment:
p2;
extension의 경우 되도록 직접 구현한것들이 아니라면 지양하고, 이전에 것들을 deprecated할 의향이 있습니다. 어떠한 이유로 해당 extension을 썼으며 아래 함수들은 어떤 방식으로 작동하는지 여쭤보고 싶어요! autolayout과 maskToBounds로 해결될 수 있다고 생각합니다.

Reply:
넵, 이건 UI에 보여질 이미지 사이즈를 조절하는게 아니라, 파베에 저장될 이미지의 크기를 조절하는 것입니다. 
이미지 크기가 너무 크면 저장되는 파일의 크기가 너무 커져, 데이터를 저장하고 읽어들이는 시간이 더 오래걸리게 되는 것을 방지하기 위해 추가한 코드입니다. 
해상도에 대해서 아주 잘 알지는 못하지만 보통 480p, 720p, 1080p, 이런식으로 해상도를 조절하는데, 
위의 코드는 더 긴 이미지 사이즈를 최대 480으로 해주어 480p 해상도보다 높지 않게 이미지 자체의 사이즈를 줄여줍니다.

<img width="200" alt="스크린샷 2022-11-23 오후 1 37 02" src="https://user-images.githubusercontent.com/80809782/203470759-132d1482-90c3-41dc-8393-a6d767a76223.png"><img width="200" alt="스크린샷 2022-11-23 오후 1 38 36" src="https://user-images.githubusercontent.com/80809782/203470731-904ecfb7-8a70-46e4-9a70-0268b396d549.png">

각각 해당 익스텐션을 사용하지 않은것과 사용한 것으로 저장되는 이미지 사이즈가 4배정도 차이나는것을 확인할 수 있습니다.

Reply:
안녕하세요 제로, apple first party에서 사진의 압축을 지원하는 기능들이 있습니다. 
https://developer.apple.com/documentation/uikit/uiimage/1624115-jpegdata을 참고하길 바래요! 
제가 개인적으로 여쭤봤던 것은 동작원리가 불분명한 코드를 지향하고자 동작원리를 알고 싶어서 여쭤본거였어요~!

Reply:
https://designcode.io/swiftui-advanced-handbook-compress-a-uiimage
아까 말씀드렸던 제가 참고한 레퍼런스입니다!
그리고 말했던것처럼 압축퀄리티를 0으로 해도 사이즈자체를 조절해 저장한것보다 3배정도 커서요, 데이터가 다 서버를 차지하는것이니 줄이면 줄일수록 좋다는 의견입니다~

Reply:
레퍼런스보다는 제로가 해당 코드의 동작원리를 알고 사용하셨는지 여쭤보고 싶습니다~~! 해당 사이트에서 사용했다고 해서 그걸 믿고 우리의 코드에 적용하는걸 원하지 않는거 같아요.. 테스트 결과 30KB, 300KB로 꽤 많이 줄여주는 건 확인했습니다. 
또한 해당 reference에서는 UIGraphicsImageRenderer를 사용하여 이미지 렌더링을 다시 했었고, 이는 우리 UIImage+ Extension에 존재하는 resize가 담당하고 있습니다. 해당 코드는 저또한 동작원리를 충분히 이해할수 있어서 쓰는데 납득이 갈거 같습니다. 혹시 resize로 바꾸시는 건 어려울까요? 
```suggestion
    var scaledToSafeUploadSize: UIImage? {
        let maxImageSideLength: CGFloat = 480
        let largerSide: CGFloat = max(size.width, size.height)
        let ratioScale: CGFloat = largerSide > maxImageSideLength ? largerSide / maxImageSideLength : 1
        let newImageSize = CGSize(width: size.width / ratioScale, height: size.height / ratioScale)
        
        return self.resize(to: newImageSize)
    }
    
    // 이미지를 입력받은 size에 맞춰 조절
    private func image(scaledTo size: CGSize) -> UIImage? {
        defer {
            UIGraphicsEndImageContext()
        }
        
        UIGraphicsBeginImageContextWithOptions(size, true, 0)
        draw(in: CGRect(origin: .zero, size: size))
        
        return UIGraphicsGetImageFromCurrentImageContext()
    }
```

Reply:
네 저는 이해했다고 판단하고 작성한 코드가 맞습니다, 어느부분에서 제가 이해를 못했다고 생각하시는걸까요~?
그리고 어느정도 이해를 한뒤에 저희 프로젝트에 도움이 되는 레퍼런스는 응용하는것이 나쁠게 없다고 생각하는데..😂 
직접 작성하지 않은 코드를 지양하더라도 저는 데이터 크기를 줄이는 것이 필요하다 생각해 이해하고 사용한 코드입니다! 이해가 안가시는 부분이 있다면 구체적으로 질문해주세요
그리고 resize가 저는 보이지 않아서요, 같은 역할을 해주는 resize가 있다면 그걸로 바꾸도록 하겠습니담 어디있는거죠!? 기본으로 제공해주는건 아닌거 같운데..

Reply:
아무래도, 처음에는 작성 코드와 레퍼런스에서 그대로 가져온 코드가 있다보니, 전체적인 코드 흐름이나 사용목적을 전반적으로 듣고 싶었는데 다른 질문에 가려졌던거 같네여.UIImage+Extension에 데이크가 작성해둔게 있어서 참고하시길 바래요~!

Reply:
q;
defer를 쓰신 목적이 무엇인지 궁금합니다. defer를 통해서 반드시 함수의 마지막에 실행되어야 하는 함수일까요?
defer키워드를 쓰는 곳이 생소하다 보니 궁금합니다.
q;
제로가 구현한 UIGraphicsBeginImageContextWithOptions도 이미지 리사이징이 가능하고, UIImageRenderer이라는 비교적 iOS10부터 지원되는 이미지 리사이징 기능이 추가되었습니다!
해당 기능에 대한 장점을 말하자면 기존 UIGraphicsBeginImageContextWithOptions는 UIImage를 srgb 영역에 맞추어서 redering하고 해당 크기는 pixel당 4byte입니다. iOS에서 지원하는 image format은 한 pixel당 1byte부터 8byte까지 다양하며, 이를 automatically하게 적용시켜주는 기능을 위해 등장한게 UIGraphicsImageRenderer입니다. 따라서 메모리를 최적화해서 image를 rendering하는 UIGraphicsRenderer를 사용한 UIimage+Extension의 resize를 쓰면 더욱 좋을거 같습니다~!
WWDC에서도 UIGraphicsBeginImageContextWithOptions를 사용하지 말고, UIGraphicsImageRenderer를 권장하고 있습니다. 감사합니당

## 출처 
https://developer.apple.com/videos/play/wwdc2018/416

Reply:
defer는 코드 실행 뒤 마지막에 꼭 실행시켜줘야 하는 함수는 저렇게 쓰는게 좋은것 같아서 저렇게 작성해보았습니다,!
resize를 사용해 적용했습니다~!!
[eef9b65](https://github.com/DeveloperAcademy-POSTECH/MacC-Team-Vegeting/pull/141/commits/eef9b65fa9826bf690d8d991d872369893da7b17)